src/combinatorics/simple_graph/ends/defs.lean

```diff
@@ -0,0 +1,280 @@
+begin
+  split,
+  { rintro ⟨⟩, refl, },
+  { refine connected_component.ind₂ _ C D,
+    rintros ⟨v, hv⟩ ⟨w, hw⟩ h,
+    simp only [set.ext_iff, connected_component.eq, set.mem_compl_iff, set.mem_set_of_eq] at h ⊢,
+    exact ((h v).mp ⟨hv, reachable.refl _⟩).some_spec, }
+end
+
+instance : set_like (G.comp_out K) V :=
+{ coe := comp_out.supp,
+  coe_injective' := λ C D, (comp_out.eq_iff_supp_eq _ _).mpr, }
+
+namespace comp_out
+
+@[simp] lemma mem_supp_iff {v : V} {C : comp_out G K} :
+  v ∈ C ↔ ∃ (vK : v ∈ K ᶜ), connected_component_mk (G.out K) ⟨v, vK⟩ = C := iff.rfl
+
+/-- The connected component of `v`. -/
+@[reducible] def of_vertex (G : simple_graph V) {v : V} (vK : v ∈ K ᶜ) : G.comp_out K :=
+  connected_component_mk (G.out K) ⟨v, vK⟩
+
+lemma of_vertex_mem (G : simple_graph V) {v : V} (vK : v ∈ K ᶜ) :
+  v ∈ (comp_out.of_vertex G vK : G.comp_out K) := ⟨vK, rfl⟩
+
+lemma of_vertex_eq_of_adj (G : simple_graph V) {v w : V} (vK : v ∈ K ᶜ) (wK : w ∈ K ᶜ) :
+  G.adj v w → comp_out.of_vertex G vK = comp_out.of_vertex G wK :=
+by { rw [connected_component.eq], rintro a, apply adj.reachable, exact a }
+
+/-- The induced subgraph of `V` given by the vertices of `C`. -/
+@[reducible, protected]
+def subgraph (C : comp_out G K) : G.subgraph := (⊤ : G.subgraph).induce (C : set V)
+
+/-- The infinite connected components. -/
+@[reducible]
+def inf (C : G.comp_out K) := (C : set V).infinite
+
+/-- The finite connected components -/
+@[reducible, protected]
+def fin (C : G.comp_out K) := (C : set V).finite
+
+lemma coe_inj {C D : G.comp_out K} : (C : set V) = (D : set V) ↔ C = D := set_like.coe_set_eq
+
+@[simp] lemma nempty (C : G.comp_out K) : (C : set V).nonempty :=
+begin
+  refine C.ind (λ v, _),
+  use v,
+  simp only [set.mem_compl_iff, set_like.mem_coe, mem_supp_iff, connected_component.eq,
+             subtype.coe_eta, exists_prop],
+  exact ⟨v.prop, reachable.refl _⟩,
+end
+
+@[protected]
+lemma outside (C : G.comp_out K) : disjoint K C :=
+begin
+  rw set.disjoint_iff,
+  rintro v ⟨vK, vC⟩,
+  simp only [mem_supp_iff, set.mem_compl_iff, set_like.mem_coe] at vC,
+  exact vC.some vK,
+end
+
+lemma not_mem_of_mem {C : G.comp_out K} {c : V} (cC : c ∈ C) : c ∉ K :=
+λ cK, set.disjoint_iff.mp C.outside ⟨cK, cC⟩
+
+@[protected]
+lemma disjoint (C D : G.comp_out K) (ne : C ≠ D) : disjoint (C : set V) (D : set V) :=
+begin
+  rw set.disjoint_iff,
+  rintro u ⟨uC, uD⟩,
+  simp only [set.mem_compl_iff, set_like.mem_coe, mem_supp_iff] at uC uD,
+  exact ne (uC.some_spec.symm.trans uD.some_spec),
+end
+
+lemma eq_of_not_disjoint (C D : G.comp_out K) (nd : ¬ disjoint (C : set V) (D : set V)) : C = D :=
+begin
+  rw set.not_disjoint_iff at nd,
+  simp only [set_like.mem_coe, mem_supp_iff] at nd,
+  obtain ⟨x, ⟨_, rfl⟩, ⟨_, rfl⟩⟩ := nd, refl,
+end
+
+/--
+No vertex of a component `C` outside of `K` is adjacent to a vertex that is
+neither in `C` nor in `K`.
+-/
+lemma nonadj : ∀ (C : G.comp_out K), ¬ (∃ (c d : V), c ∈ C ∧ d ∉ C ∧ d ∉ K ∧ G.adj c d) :=
```

Comment:
This is too ambiguous of a name.

Reply:
It's probably just how I think about it, but can be changed if there is an argument in favour of \forall.

Reply:
OK, I've got a better name and removed the `\not \exists` and it's better like this!